pkg/query-service/app/logs/v3/enrich_query_test.go

```diff
@@ -456,6 +456,129 @@ func TestJsonEnrich(t *testing.T) {
 	}
 }
 
+var TestJsonReplaceFieldData = []struct {
+	Name   string
+	Filter v3.FilterItem
+	Result v3.FilterItem
+}{
+	{
+		Name: "key in nested json",
+		Filter: v3.FilterItem{
+			Key: v3.AttributeKey{
+				Key:      "body.method.name",
+				DataType: v3.AttributeKeyDataTypeString,
+				Type:     v3.AttributeKeyTypeUnspecified,
+			},
+			Operator: "has",
+			Value:    "index_service",
+		},
+		Result: v3.FilterItem{
+			Key: v3.AttributeKey{
+				Key:      "method.name",
+				DataType: v3.AttributeKeyDataTypeString,
+				Type:     v3.AttributeKeyTypeTag,
+				IsJSON:   false,
+			},
+			Operator: "has",
+			Value:    "index_service",
+		},
+	},
+	{
+		Name: "key at top level",
+		Filter: v3.FilterItem{
+			Key: v3.AttributeKey{
+				Key:      "body.status",
+				DataType: v3.AttributeKeyDataTypeInt64,
+				Type:     v3.AttributeKeyTypeUnspecified,
+			},
+			Operator: "=",
+			Value:    10,
+		},
+		Result: v3.FilterItem{
+			Key: v3.AttributeKey{
+				Key:      "status",
+				DataType: v3.AttributeKeyDataTypeInt64,
+				Type:     v3.AttributeKeyTypeTag,
+				IsJSON:   false,
+			},
+			Operator: "=",
+			Value:    10,
+		},
+	},
+	{
+		Name: "key not present",
+		Filter: v3.FilterItem{
+			Key: v3.AttributeKey{
+				Key:      "body.status.code",
+				DataType: v3.AttributeKeyDataTypeInt64,
+				Type:     v3.AttributeKeyTypeUnspecified,
+			},
+			Operator: "=",
+			Value:    10,
+		},
+		Result: v3.FilterItem{
+			Key: v3.AttributeKey{
+				Key:      "body.status.code",
+				DataType: v3.AttributeKeyDataTypeInt64,
+				Type:     v3.AttributeKeyTypeUnspecified,
+				IsJSON:   false,
+			},
+			Operator: "=",
+			Value:    10,
+		},
+	},
+	{
+		Name: "key materialized",
+		Filter: v3.FilterItem{
+			Key: v3.AttributeKey{
+				Key:      "body.data.error",
+				DataType: v3.AttributeKeyDataTypeString,
+				Type:     v3.AttributeKeyTypeUnspecified,
+			},
+			Operator: "=",
+			Value:    10,
+		},
+		Result: v3.FilterItem{
+			Key: v3.AttributeKey{
+				Key:      "data.error",
+				DataType: v3.AttributeKeyDataTypeString,
+				Type:     v3.AttributeKeyTypeTag,
+				IsJSON:   false,
+				IsColumn: true,
+			},
+			Operator: "=",
+			Value:    10,
+		},
+	},
+}
+
+func TestJsonReplaceField(t *testing.T) {
+	fields := map[string]v3.AttributeKey{
+		"method.name": {
+			Key:      "method.name",
+			DataType: v3.AttributeKeyDataTypeString,
+			Type:     v3.AttributeKeyTypeTag,
+		},
+		"status": {
+			Key:      "status",
+			DataType: v3.AttributeKeyDataTypeInt64,
+			Type:     v3.AttributeKeyTypeTag,
+		},
+		"data.error": {
```

Comment:
yes all the existing columns were moved to new format i.e `$$` .

`GetClickhouseColumnName` function converts the `.` to `$$` during query building process.

Reply:
What is this piece of code doing?
```
	if field, ok := fields[key]; ok && field.DataType == filter.Key.DataType {
		filter.Key = field
	}

	return filter
```

It checks if the key exists in the fields but it would never exist in the fields because there are no columns with dot anymore?

Reply:
so the fields are fetch from the attributes tables and checked if they are selected or not using the `extractSelectedAndInterestingFields` . So in this case it will work. 

so if a field name is `method.name` and there is a `attribute_string_method$$name` column then method.name is selected.

Now once someone seraches for `body.method.name = xyz` then it will be mapped to `method.name` selected field, which will be converted to `attribute_string_method$$name` during query

Reply:
Where does it happen that `attribute_string_method$$name` becomes a `method.name` key in the `fields` map?

Reply:
* so first we get all the fields (without $) from the attributes table.
* and then for each field we check if the field is selected or not .

it's the `GetLogFields` function

Reply:
This never happens i.e `attribute_string_method$$name to method.name` , it's always `method.name to attribute_string_method$$name` during the query time if the field `isColumn` is true.

when we look `method.name` . 
There will be a field name 
```
{
 "key": "method.name",
 "type": "attribute",
 "dataType": "string",
 "isColumn":  true
}
```

If the `isColumn` is true then during query it is transformed to `attribute_string_method$$name` else the corresponding attribute in the array is used.

Reply:
Please bear with me while I understand this change. The shared snippet above is from `enrichLogsQuery` which is invoked in `Enrich` which is getting called in queryRangeX function. The queryRange gets the `fields` from `getLogFieldsV3` which in turn gets from `GetLogFields`. 


For this condition `if field, ok := fields[key]; ok ` to evaluate to true there must be a key with dots in it? And if the `attribute_string_method$$name ` never happens then there will never be `method.name` inside `fields` map. What is the point of `jsonReplaceField`?

Reply:
No worries,
so this is what happens.
my filter is
```
{
    "key": {
        "key": "body.method.name",
        "dataType": "string",
        "type": "",
    },
    "op": "=",
    "value": "GET"
}
```
now if there is a field name `method.name` already extracted using parsers then 
the fields map will contain
```
{
 "key": "method.name",
 "type": "attribute",
 "dataType": "string",
 "isColumn":  true
}
```

Now the filter will be updated to 
```
{
    "key": {
         "key": "method.name",
         "type": "attribute",
         "dataType": "string",
         "isColumn":  true
     },
    "op": "=",
    "value": "GET"
}
```

Now during the query building process, if the `isColumn` is true then `attribute_string_method$$name` will be used.

Reply:
This is true only if there is a `method.name` that is part of attributes (and the attributes table) from the writing side. Is it always guaranteed to be part of the attributes? Don't we also allow materializing from the JSON body Or is it something yet to be done?

Reply:
Yeah right now the only way to create a materialized column is through attributes only.

as of now that is gauranteed.